datafusion/functions/src/datetime/date_trunc.rs

```diff
@@ -123,7 +123,9 @@ impl ScalarUDFImpl for DateTruncFunc {
 
     fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
         match &arg_types[1] {
-            Timestamp(Nanosecond, None) | Utf8 | Null => Ok(Timestamp(Nanosecond, None)),
+            Timestamp(Nanosecond, None) | Utf8 | DataType::Date32 | Null => {
```

Comment:
I also tried to support `Date64` but got many type coercion issues between `Timestamp` and `Date64`. I think we should handle them in another PR if needed.